### PR TITLE
PVT: Compute HDOP and VDOP correctly

### PR DESCRIPTION
**Pull request by henryhallam**
We were doing two things wrong in our calculations of HDOP and VDOP:
1. Mistakenly squaring the diagonal elements of H_ned before square rooting
2. Rotating only the main diagonal of H to form H_ned.  This isn't legit, because H isn't orthonormal[1].  You have to transform the whole matrix, then look at the diagonal of the result.

This is verified to match SimGEN's idea of all the DOPs to within 1% (remainder probably explained by different original convariance matrices, plus perhaps different representations of NED frame).

cc @fnoble @imh @kovach 

[1] Did I use that linear algebra vocabulary correctly? I mean, the covariance is ellipsoidal, not spherical.

### CODE REVIEW

**Comment by henryhallam** [comment]
I'm gonna refactor this a little to make use of some more of @peddie's handy-dandy linear algebra functions

**Comment by henryhallam** [comment]
@kovach @fnoble ready for review

**Comment by imh** [comment]
re: the comment about terminology. It's usually called an orthogonal matrix, not orthonormal. Its columns form an orthonormal basis.

File: src/pvt.c [hunk 1]
```diff
@@ -67,33 +67,28 @@ static void compute_dops(const double H[4][4],
...
-  wgsecef2ned(H_pos_diag, pos_ecef, H_ned);
-  dops->vdop = sqrt(H_ned[2]*H_ned[2]);
-  dops->hdop = sqrt(H_ned[0]*H_ned[0] + H_ned[1]*H_ned[1]);
+  dops->gdop = sqrt(pdop_sq + H[3][3]);
+
+  /* HDOP and VDOP are Horizontal and Vertical; we need to rotate H
+   * into NED frame and then take the separate components. */
+  double M[3][3], Mt[3][3];
+  double H_3x3[3][3], HM[3][3], H_ned[3][3];
+  /* TODO: save a little CPU by taking advantage of symmetry */
```
Comment (imh) [inline src/pvt.c]: you could also save more CPU by calculating HDOP from VDOP and PDOP:
For v = ecef2n_vector, VDOP^2 = v * H * v^T, which is easy to compute cheaply. Then use HDOP = sqrt(PDOP^2 - VDOP^2). It's probably numerically stable enough too for the values we'll see.

File: tests/check_pvt.c [hunk 2]
```diff
@@ -109,6 +109,42 @@ START_TEST(test_disable_pvt_raim)
 }
 END_TEST
 
+START_TEST(test_dops)
+{
+  u8 n_used = 6;
+  gnss_solution soln;
+  dops_t dops = {.pdop = 22, .gdop = 22, .tdop = 22, .hdop = 22, .vdop = 22};
```
Comment (imh) [inline tests/check_pvt.c]: why does this need to be initialized?

Comment (henryhallam) [inline tests/check_pvt.c]: in case calc_PVT fails to set it.

File: src/pvt.c (hunk 1 above)
Comment (henryhallam) [inline src/pvt.c]: do you mean: 
 - Don't form M, and don't rotate H into NED frame
 - Instead, rotate the North unit vector into ECEF frame
 - smash that through H to form VDOP (technical term)
 - Find HDOP as sqrt(PDOP^2 - VDOP^2)
..?

Comment (imh) [inline src/pvt.c]: Oops, i meant the down component, not the north one. Use that to project out the up/down variance from H, which gives VDOP, then find HDOP from there.

Comment (henryhallam) [inline src/pvt.c]: heh yeah, I meant Down too.  OK, that's a good idea, I'll give it a try.  Now we have unit tests, it's a snap :)

**Pull request merged**